### PR TITLE
posix: signal: include_next to get the libc signal() decl

### LINKED ISSUE #56294: Using `signals` from `libc` is broken
**Describe the bug**
An application with `CONFIG_POSIX_API` which was working before the [commit](https://github.com/zephyrproject-rtos/zephyr/commit/43110ddb71680db3ddd3411317dbaf0476e298f5) is now broken.

Zephyr's POSIX implementation [doesn't support ](https://docs.zephyrproject.org/latest/services/portability/posix.html#posix-signals)`signal`. But `libc` does support it, but after the change when `CONFIG_POSIX_API` is enabled, and if we want to use POSIX `signals` then its broken as in the list of include directories `include/zephyr/posix/signal.h` appears first before `libc` headers and that file doesn't has `signal` related definitions. (It only has definitions needed to support pthread_cond_signal).

**To Reproduce**
Steps to reproduce the behavior: Clone `https://github.com/krish2718/zephyr/tree/libc_signal` and build `samples/net/wifi` (I have used VS code to build.

**Expected behavior**
Successful build

**Impact**
Existing application (downstream) is broken.

**Logs and console output**
```
FAILED: CMakeFiles/app.dir/src/wifi_test.c.obj 
ccache /home/tach/work/zephyr-sdk-0.15.2/arm-zephyr-eabi/bin/arm-zephyr-eabi-gcc -DKERNEL -DMBEDTLS_CONFIG_FILE=\"config-tls-generic.h\" -DNRF5340_XXAA_APPLICATION -DNRF_SKIP_FICR_NS_COPY_TO_RAM -D_FORTIFY_SOURCE=2 -D__LINUX_ERRNO_EXTENSIONS__ -D__PROGRAM_START -D__ZEPHYR__=1 -I/home/tach/work/upstream/zephyr/include/zephyr -I/home/tach/work/upstream/zephyr/include -I/home/tach/work/upstream/zephyr/samples/net/wifi/build/zephyr/include/generated -I/home/tach/work/upstream/zephyr/soc/arm/nordic_nrf/nrf53 -I/home/tach/work/upstream/zephyr/lib/libc/newlib/include -I/home/tach/work/upstream/zephyr/lib/util/fnmatch/. -I/home/tach/work/upstream/zephyr/soc/arm/nordic_nrf/common/. -I/home/tach/work/upstream/modules/hal/cmsis/CMSIS/Core/Include -I/home/tach/work/upstream/modules/hal/atmel/include -I/home/tach/work/upstream/modules/hal/nordic/nrfx -I/home/tach/work/upstream/modules/hal/nordic/nrfx/drivers/include -I/home/tach/work/upstream/modules/hal/nordic/nrfx/mdk -I/home/tach/work/upstream/zephyr/modules/hal_nordic/nrfx/. -I/home/tach/work/upstream/zephyr/include/zephyr/posix -I/home/tach/work/upstream/modules/crypto/mbedtls/include -I/home/tach/work/upstream/zephyr/modules/mbedtls/configs -Os -imacros /home/tach/work/upstream/zephyr/samples/net/wifi/build/zephyr/include/generated/autoconf.h -ffreestanding -fno-common -g -gdwarf-4 -fdiagnostics-color=always -mcpu=cortex-m33 -mthumb -mabi=aapcs -mfp16-format=ieee --sysroot=/home/tach/work/zephyr-sdk-0.15.2/arm-zephyr-eabi/arm-zephyr-eabi -imacros /home/tach/work/upstream/zephyr/include/zephyr/toolchain/zephyr_stdint.h -Wall -Wformat -Wformat-security -Wno-format-zero-length -Wno-main -Wno-pointer-sign -Wpointer-arith -Wexpansion-to-defined -Wno-unused-but-set-variable -Werror=implicit-int -fno-asynchronous-unwind-tables -fno-pie -fno-pic -fno-reorder-functions -fno-defer-pop -fmacro-prefix-map=/home/tach/work/upstream/zephyr/samples/net/wifi=CMAKE_SOURCE_DIR -fmacro-prefix-map=/home/tach/work/upstream/zephyr=ZEPHYR_BASE -fmacro-prefix-map=/home/tach/work/upstream=WEST_TOPDIR -ffunction-sections -fdata-sections -specs=nano.specs -std=c99 -MD -MT CMakeFiles/app.dir/src/wifi_test.c.obj -MF CMakeFiles/app.dir/src/wifi_test.c.obj.d -o CMakeFiles/app.dir/src/wifi_test.c.obj -c /home/tach/work/upstream/zephyr/samples/net/wifi/src/wifi_test.c
/home/tach/work/upstream/zephyr/samples/net/wifi/src/wifi_test.c: In function 'main':
/home/tach/work/upstream/zephyr/samples/net/wifi/src/wifi_test.c:12:9: warning: implicit declaration of function 'signal' [-Wimplicit-function-declaration]
   12 |         signal(SIGINT, NULL);
      |         ^~~~~~
/home/tach/work/upstream/zephyr/samples/net/wifi/src/wifi_test.c:12:16: error: 'SIGINT' undeclared (first use in this function)
   12 |         signal(SIGINT, NULL);
      |                ^~~~~~
/home/tach/work/upstream/zephyr/samples/net/wifi/src/wifi_test.c:12:16: note: each undeclared identifier is reported only once for each function it appears in
```

**Environment (please complete the following information):**
 - OS: LInux
 - Toolchain; Zephyr SDK 0.15.2
 - Commit SHA: `https://github.com/krish2718/zephyr/tree/libc_signal` 

**Additional context**
We could solve this in many ways:

1. Add `signal` support properly to Zephyr POSIX implementation (long term)
2. Tweak Cmake to include `libc` headers before Zephyr, if its possible, but I don't see `prepend_include_directories` in cmake.
3. As the commit log says, avoid using `CONFIG_POSIX_API` and instead individually enable POSIX_* modules as a hack.

I am leaning towards trying out #3 or #2 for short term.

@cfriedt  @stephanosio @sachinthegreen

### PR DESCRIPTION
**Pull request by cfriedt**
**Note: Coding Guidellines issue is a false positive**

Although the ANSI C version of signal() is not yet supported in Zephyr, because Zephyr does not use process-wide signals, some users rely on its declaration.

Include the libc version of signal.h when POSIX signals are not configured.

Also fixed some inadvertent typing to be more compliant with the spec.

Fixes #56294

### CODE REVIEW

**Comment by cfriedt** [comment]
Unrelated errors are unrelated.

**Review by ycsin — DISMISSED** [review]
edited: please ignore this, overlooked some of the tests

File: include/zephyr/posix/signal.h [hunk 1]
```diff
@@ -77,23 +75,31 @@ int sigismember(const sigset_t *set, int signo);
...
+typedef void (*sighandler_t)(int signo);
+sighandler_t signal(int signum, sighandler_t handler);
 
 #ifdef __cplusplus
 }
 #endif
 
+#else /* CONFIG_POSIX_SIGNAL */
+
+#include_next <signal.h>
```
Comment (keith-packard) [inline include/zephyr/posix/signal.h:101]: Why is this needed? Does `-Iinclude/zephyr/posix` get added to the compiler command line?

File: lib/libc/minimal/include/signal.h [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+typedef int	sig_atomic_t;		/* Atomic entity type (ANSI) */
+
+/* Note: sighandler_t is a gnu-ism, but it simplifies the declaration below */
+typedef void (*sighandler_t)(int signo);
+
+sighandler_t signal(int signum, sighandler_t handler);
```
Comment (keith-packard) [inline lib/libc/minimal/include/signal.h:19]: Does minimal C library support this function? If not, you probably shouldn't declare it here.

File: include/zephyr/posix/signal.h (hunk 1 above)
Comment (cfriedt) [reply include/zephyr/posix/signal.h:101]: > Why is this needed? Does `-Iinclude/zephyr/posix` get added to the compiler command line?

This one is actually to include the next signal.h in the standard include path, which would typically be the libc

File: lib/libc/minimal/include/signal.h (hunk 2 above)
Comment (keith-packard) [reply lib/libc/minimal/include/signal.h:19]: The C standard specifies behavior for the `raise()` and `abort()` functions related to `signal()`, so there are bits which "should" be provided as part of a C environment. (whether any of that is useful is another question). However, if you want to just get rid of the whole thing and pretend signals don't exist, then I'd suggest you might be better served by ignoring calls to signal rather than reporting an error. In this case, you're promising to call the specified function whenever the related signal is raised. As said signal will *never* be raised (unless the application uses `raise()` itself, then returning SIG_DFL would prevent applications that check the return value from failing while still meeting the API promise.

Comment (cfriedt) [reply lib/libc/minimal/include/signal.h:23]: Fixed! Should pass CI now, since unrelated errors are fixed now.

**Comment by cfriedt** [comment]
This is getting annoying. I might trim it down to the bare essentials.

**Comment by cfriedt** [comment]
Closing in favour of the other signal.h PR
